File: crates/openduo-tools/src/registry.rs
use crate::cicd::CicdTools;
use crate::issues::IssuesTools;
use crate::labels::LabelTools;
use crate::merge_requests::MergeRequestTools;
use crate::milestones::MilestoneTools;
use crate::pipelines::PipelineTools;
use crate::projects::ProjectTools;
use crate::repositories::RepositoryTools;
use crate::users::UserTools;
use anyhow::Result;
use async_trait::async_trait;
use openduo_core::{config::Config, gitlab_client::GitLabClient, types::ToolDefinition};
use std::collections::HashMap;

#[async_trait]
pub trait Tool: Send + Sync {
    fn name(&self) -> &str;
    fn description(&self) -> &str;
    fn parameters_schema(&self) -> serde_json::Value;
    async fn execute(&self, args: serde_json::Value) -> Result<String>;

    fn definition(&self) -> ToolDefinition {
        ToolDefinition {
            name: self.name().to_string(),
            description: self.description().to_string(),
            parameters: self.parameters_schema(),
        }
    }
}

pub struct ToolRegistry {
    tools: HashMap<String, Box<dyn Tool>>,
}

impl ToolRegistry {
    pub fn new(config: Config) -> Self {
        let client = GitLabClient::new(config);
        let mut tools: HashMap<String, Box<dyn Tool>> = HashMap::new();

        for tool in IssuesTools::all(client.clone()) {
            tools.insert(tool.name().to_string(), tool);
        }
        for tool in MergeRequestTools::all(client.clone()) {
            tools.insert(tool.name().to_string(), tool);
        }
        for tool in PipelineTools::all(client.clone()) {
            tools.insert(tool.name().to_string(), tool);
        }
        for tool in RepositoryTools::all(client.clone()) {
            tools.insert(tool.name().to_string(), tool);
        }
        for tool in ProjectTools::all(client.clone()) {
            tools.insert(tool.name().to_string(), tool);
        }
        for tool in UserTools::all(client.clone()) {
            tools.insert(tool.name().to_string(), tool);
        }
        for tool in CicdTools::all(client.clone()) {
            tools.insert(tool.name().to_string(), tool);
        }
        for tool in MilestoneTools::all(client.clone()) {
            tools.insert(tool.name().to_string(), tool);
        }
        for tool in LabelTools::all(client.clone()) {
            tools.insert(tool.name().to_string(), tool);
        }

        Self { tools }
    }

    pub fn definitions(&self) -> Vec<ToolDefinition> {
        self.tools.values().map(|t| t.definition()).collect()
    }

    pub async fn execute(&self, name: &str, args: serde_json::Value) -> Result<String> {
        let span = tracing::info_span!("tool_execute", tool_name = %name);
        let _enter = span.enter();
        tracing::info!(tool = %name, args = %args, "Tool invocation");
        match self.tools.get(name) {
            Some(tool) => {
                let result = tool.execute(args).await;
                match &result {
                    Ok(r) => tracing::info!(tool = %name, result_len = r.len(), "Tool success"),
                    Err(e) => tracing::error!(tool = %name, error = %e, "Tool failed"),
                }
                result
            }
            None => anyhow::bail!("Unknown tool: {}", name),
        }
    }
}

File: extension/server/tools/registry.ts
import type { Config, WorkspaceFolder } from '../config';
import { GitLabClient } from '../gitlabClient';
import type { ToolDefinition } from '../types';
import type { Tool, ToolContext } from './tool';
import { toolDefinition } from './tool';
import { issueTools } from './issues';
import { mergeRequestTools } from './mergeRequests';
import { pipelineTools } from './pipelines';
import { repositoryTools } from './repositories';
import { projectTools } from './projects';
import { userTools } from './users';
import { cicdTools } from './cicd';
import { milestoneTools } from './milestones';
import { labelTools } from './labels';
import { graphqlQueryTools } from './graphqlQueries';
import { snippetTools } from './snippets';
import { groupTools } from './groups';
import { environmentTools } from './environments';
import { wikiTools } from './wiki';
import { workspaceTools } from './workspace';
import { editorContextTools } from './editorContext';
import { diagnosticsTools } from './diagnostics';

export class ToolRegistry {
  private tools: Map<string, Tool>;
  private _activeFolder: WorkspaceFolder | undefined;
  readonly workspaceFolders: WorkspaceFolder[];

  constructor(config: Config) {
    const client = new GitLabClient(config);
    this.tools = new Map();
    this.workspaceFolders = config.workspaceFolders;

    // Default to the first workspace folder if available
    this._activeFolder = config.workspaceFolders[0];

    const allTools = [
      // REST API tools
      ...issueTools(client),
      ...mergeRequestTools(client),
      ...pipelineTools(client),
      ...repositoryTools(client),
      ...projectTools(client),
      ...userTools(client),
      ...cicdTools(client),
      ...milestoneTools(client),
      ...labelTools(client),
      ...snippetTools(client),
      ...groupTools(client),
      ...environmentTools(client),
      ...wikiTools(client),
      // GraphQL API tools (richer queries)
      ...graphqlQueryTools(client),
      // Local workspace tools (read, list, search local files)
      ...workspaceTools(
        config.workspaceFolders,
        () => this._activeFolder,
      ),
      // Editor context tools (IPC to extension host)
      ...editorContextTools(),
      // Diagnostics tools (IPC to extension host)
      ...diagnosticsTools(),
    ];

    for (const tool of allTools) {
      this.tools.set(tool.name, tool);
    }
  }

  get activeFolder(): WorkspaceFolder | undefined {
    return this._activeFolder;
  }

  setActiveFolder(name: string): boolean {
    const folder = this.workspaceFolders.find(f => f.name === name);
    if (folder) {
      this._activeFolder = folder;
      return true;
    }
    return false;
  }

  definitions(): ToolDefinition[] {
    return Array.from(this.tools.values()).map(toolDefinition);
  }

  async execute(name: string, args: Record<string, unknown>, context?: ToolContext): Promise<string> {
    const tool = this.tools.get(name);
    if (!tool) throw new Error(`Unknown tool: ${name}`);
    console.log(`[tool] Invoking ${name} with args: ${JSON.stringify(args)}`);
    const result = await tool.execute(args, context);
    console.log(`[tool] ${name} returned ${result.length} chars`);
    return result;
  }
}
